libraries/plugins/debug_node/plugin.cpp

```diff
@@ -0,0 +1,671 @@
+    auto args_vector = *(args.args);
+
+    src_filename = args_vector[0].as_string();
+    count = args_vector[1].as_int64();        
+    if (args_count > 2) {
+        skip_validate_invariants = args_vector[3].as_bool();
+    }
+
+    auto &db = my->database();
+    return db.with_read_lock([&]() {
+        return my->debug_push_blocks( src_filename, count, skip_validate_invariants );
+    });
+}
+
+DEFINE_API ( plugin, debug_generate_blocks_until ) {
+    std::string debug_key;
+    fc::time_point_sec head_block_time;
+    bool generate_sparsely = true;
+
+    FC_ASSERT(args.args.valid(), "Invalid parameters" ) ;
+
+    auto args_count = args.args->size() ;
+
+    FC_ASSERT( args_count > 0 && args_count < 5, "Wrong parameters number, given ${n}", ("n", args_count) );
+    auto args_vector = *(args.args);
+
+    debug_key = args_vector[0].as_string();
+    head_block_time = fc::time_point_sec::from_iso_string( args_vector[1].as_string() );      
+
+    if (args_count > 2) {
+        generate_sparsely = args_vector[3].as_bool();
+    }
+
+    auto &db = my->database();
+    return db.with_read_lock([&]() {
+        return my->debug_generate_blocks_until( debug_key, head_block_time, generate_sparsely );
+    });
+}
+
+DEFINE_API ( plugin, debug_pop_block ) {
+    auto &db = my->database();
+    return db.with_read_lock([&]() {
+        return my->debug_pop_block();
+    });
+}
+
+DEFINE_API ( plugin, debug_get_witness_schedule ) {
+    auto &db = my->database();
+    return db.with_read_lock([&]() {
+        return my->debug_get_witness_schedule();
+    });
+}
+
+// DEFINE_API ( plugin, debug_get_hardfork_property_object ) {
+//     auto tmp = args.args->at(0).as<debug_get_hardfork_property_object_a>();
+//     auto &db = my->database();
+//     return db.with_read_lock([&]() {
+//         return my->debug_get_hardfork_property_object(tmp);
+//     });
+// }
+
+DEFINE_API ( plugin, debug_set_hardfork ) {
+    uint32_t hardfork_id;
+
+    FC_ASSERT(args.args.valid(), "Invalid parameters" ) ;
+
+    auto args_count = args.args->size() ;
+
+    FC_ASSERT( args_count == 1, "Wrong parameters number, given ${n}", ("n", args_count) );
+    auto args_vector = *(args.args);
+    hardfork_id = args_vector[0].as_int64();
+
+    auto &db = my->database();
+    return db.with_read_lock([&]() {
+        my->debug_set_hardfork( hardfork_id );
+        return void_type() ;
+    });
+}
+
+DEFINE_API ( plugin, debug_has_hardfork ) {
+    uint32_t hardfork_id;
+
+    FC_ASSERT(args.args.valid(), "Invalid parameters" ) ;
+
+    auto args_count = args.args->size() ;
+
+    FC_ASSERT( args_count == 1, "Wrong parameters number, given ${n}", ("n", args_count) );
+    auto args_vector = *(args.args);
+    hardfork_id = args_vector[0].as_int64();
+
+    auto &db = my->database();
+    return db.with_read_lock([&]() {
+        return my->debug_has_hardfork( hardfork_id );
+    });
+}
+
+
+
+/*void plugin::save_debug_updates( fc::mutable_variant_object& target )
```

Comment:
Why these lines are commented? They don't work or not?

Reply:
@afalaleev, these lines are commented in steem.